File: examples/SLC/spinlock.cpp
#include <atomic>

class SpinLock {
  std::atomic<int64_t> locked;

public:
  SpinLock(): locked(0) {
  }

  bool lock();
  void unlock();
};

__attribute__((always_inline)) inline bool SpinLock::lock() {
  int64_t v = 0;
  return locked.compare_exchange_strong(v, 1, std::memory_order_acquire, std::memory_order_relaxed);
}

__attribute__((always_inline)) inline void SpinLock::unlock() {
  locked.store(0, std::memory_order_release);
}

void thread0(SpinLock &l, int64_t X, int64_t &data) {
  for (int64_t i = 0; i < X; ++i) {
    if (l.lock()) {
      data += 42;
      l.unlock();
      break;
    }
  }
}

void thread1(SpinLock &l, int64_t X, int64_t &data) {
  for (int64_t i = 0; i < X; ++i) {
    if (l.lock()) {
      data += 42;
      l.unlock();
      break;
    }
  }
}

void thread2(SpinLock &l, int64_t X, int64_t &data) {
  for (int64_t i = 0; i < X; ++i) {
    if (l.lock()) {
      data += 42;
      l.unlock();
      break;
    }
  }
}

File: examples/SLR/spinlock.rs
use std::sync::atomic::AtomicIsize;
use std::sync::atomic::Ordering::{Acquire, Relaxed, Release};

pub struct Spinlock {
  locked: AtomicIsize,
}

impl Spinlock {
  #[inline(always)]
  pub fn lock(&self) -> bool {
    match self.locked.compare_exchange(0, 1, Acquire, Relaxed) {
      Ok(_) => true,
      Err(_) => false,
    }
  }

  #[inline(always)]
  pub fn unlock(&self) {
    self.locked.store(0, Release);
  }
}

pub fn thread0(l: &Spinlock, x: i64, data: &AtomicIsize) {
  for _i in 0..x {
    if l.lock() {
      let d = data.load(Relaxed);
      data.store(d + 42, Relaxed);
      l.unlock();
      break;
    }
  }
}

pub fn thread1(l: &Spinlock, x: i64, data: &AtomicIsize) {
  for _i in 0..x {
    if l.lock() {
      let d = data.load(Relaxed);
      data.store(d + 42, Relaxed);
      l.unlock();
      break;
    }
  }
}

pub fn thread2(l: &Spinlock, x: i64, data: &AtomicIsize) {
  for _i in 0..x {
    if l.lock() {
      let d = data.load(Relaxed);
      data.store(d + 42, Relaxed);
      l.unlock();
      break;
    }
  }
}
